---
title: "Cortexomics"
author: "Dermot Harnett"
output: github_document
---



```{r setup, include=FALSE}
#output: rmarkdown::github_document
library(knitr)
```

```{r timestamp, echo = TRUE}
Sys.time()
```


## Intro

See /fast/work/groups/ag_ohler/dharnet_m/cortexomics/text/paper_mockup.pdf

- Importance of understanding gene regulation
- Importance of understanding gene regulation at multiple levels
- Specific importance to the brain
	- complexity of translational regulation in the brain
	- importance of complexes
- Distinction between intra and inter gene variation, discussion of it's importance
- Lack of studies in living tissue that examine Gene expression at multiple levels
-  


In this work, we have carried out a comprehensive analysis of gene regulation in the developing murine neocortex, assaying gene regulation at the level of steady state RNA levels, Ribosome bound RNA, and steady state protein.

First, we demonstrate how by accounting for the unique technical biases of each assay, a more accurate account of gene regulation throughout the central dogma (intro to dogma quant here) can be achieved.

Secondly, our analysis provides a unique window into the relative magnitude and prevalence of transcriptional, translational and post translational gene regulation mechanisms over the course of development in a well studied tissue, allowing us to compare variance between and within gene's expression levels over time.

Thirdly, our analysis builds on existing literature (_ demonstrating the importance of translational regulatory mechanisms to in cortical development.





### Refs to include

[Fujii et al 2017](https://www.ncbi.nlm.nih.gov/pubmed/28195124)
FACS+Riboseq
translational regulation of lots of cell signalling molecules - Shh wnt hippo pi3k mapk
uORFs prevalent. confirm importance of uORF for Ptch1 signaling. lots of mRNAs have translational but no transcriptional reg.
	- let's look at their processing, did they do p sites? I  worry about uorfs and offsets
	- exclude the last 5 and first 15 codons... this is going to kill power, but I see where they're coming from with CHX.
	- Base a lot of theri analysis on absolute TE levels.... there are a LOT of confounders in that not sure if they accounted for all - mapping, splicing, ramping, elongation speed, stalling...

[](https://www.cell.com/cell/abstract/S0092-8674(17)30582-2)
## results
other attached packages:
[1] gdtools0.1.7 compiler==3.4.3 Rcpp==0.12.18   svglite==1.2.1 
> 



###  Figure 1 - Data Quality
- figure showing stages, data layout, absorbance profile for polysomes. EM of the protein
- pca of the data
- averagograms showing periodicity etc.
- Also comparing poly to total
- Proportions of stuff in different places
- pathways enriched in 80S mono vs POly



Rnaseq, Riboseq, Mass spec correlations, outlier robust correlations.
- Number, size, mapped reads, correlation between replicates
- Enzymes used, for RIboseq, rationale
- Likely effects this has on the data - e.g. cyclohexamine (see sup figure for sequence preference)

```{r f1 nap, out.width ='900px', out.height ='900px' ,fig.cap = "caption", dev='jpeg'}
include_graphics("plots/tmp.pdf")
```


### Figure 2 Riboseq

TE modeling - sequence motifs - cell type, neurites
Maybe circ showing shared genes in early vs. late?
"Global Characteristics"
TE timecourse
GO term sig up vs. Down


```{r f2 nap, out.width ='900px', out.height ='900px' ,fig.cap = "caption", dev='jpeg'}
include_graphics("plots/tmp.pdf")

```

Basic discussion of Riboseq
- Periodicity in our Riboseq
- Number of ORFs recovered
- New ORFS recoovered - e.g. 
- Translational Regulation Tests - Ribodiff


### Figure 3 Mass Spec Integration, clustering (jpeg file)
```{r f3, out.width ='900px', out.height ='900px' ,fig.cap = "caption", dev='jpeg'}
include_graphics("plots/tmp.jpeg")

```

 - observation about non-equilibrium kinetics
 - Observation that many genes don't have perceptible reaction to changing riboseq levels - still valid?
 - calculation of degredation ratess, motivation
 - Trajectory based clustering
 - Novel peptides in UTR?


 Chat with machmud - permutation test is one way to decide how many components or clusters
 SHould I be clustering rTE and degredtion
 Can I







### Figure 4 - targets, profiles
```{r f4, out.width ='900px', out.height ='900px' ,fig.cap = "caption", dev='jpeg'}
plot(3) 
# include_graphics("plots/tmp.jpeg")
```
 Targets.


- possible clustering methods
	- timeless
	- self organizing maps

- Sequence analysis

### Figure 5-7 - Wetlab

## Discussion
Resource for study of cortical development
For comparison to single cell data



## Supplementary Figures

## Supp Figure 1

Data quality
Duplication rates

![The multiqc report](pipeline/multiqc_report.html) shows various summary statistics for the libraries.




## Supp Figure 2
- Correlation between Riboseq Replicates
- Sequence biases in the Riboseq replicates
- Attemps to compensate?
'We leveraged replicate measurements to assess data quality and its dependence on several parameters, including alignment strategy, mRNA enrichment method, PCR artifacts, gene length normalization, and batch effects (Supplemental Fig. S2A–E; Supplemental Methods).''
- 

## Supp Figure 3

## References
To read:
Integrated Transcriptome and Proteome Analyses Reveal Organ-Specific Proteome Deterioration in Old Rats.



Javonavic et al 2015
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4506746/

Cenik et al 2015
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4617958/
 - Lots of influence on translation by variants in start codons, ORFs.
 - Inter individual variation

Yuzwa et al 2017 cell rep
Developmental Emergence of Adult Neural Stem Cells as Revealed by Single-Cell Transcriptional Profiling.
https://www.ncbi.nlm.nih.gov/pubmed/29281841


Cho et al 2015
memory and translational reg
Methods should be useful, note they measure enrichment of reads in coding sequences.

Translation in nuerons review
https://www.cell.com/action/showPdf?pii=S0896-6273%2817%2930926-1
Process of initiation - PIC forms when cap binding protein binds the 43S with meTRNA, small subunit, various elongation factors, prob also PolyA bindign protein or PABB
IRES exist  bear in mind,
Acceptor, P-site (peptidly) and exit site - psite on mTRNA is where it starts, a site awaits tRNA
tRNA arrives, GTP hydrolysis, efactor rele, peptidyl bond transfered to A site AA, conf change, hybrid state with tRNA stems shifted so they are in E-P, anticodon loops still in  P-A, e2Fbinding at GTPhydro then resets releasing esite tRNA, shifting Psite tRNA's anticodon
uORFs involved in stress response _cite_
mTOR phosphyorylates e factors eIF4G and eIF4B, and of the initia-tion factor inhibitors eIF4E-binding protein (4E-BP/EIF4EBP), this 
some neurodegen conditions associated with defective elF2 reg ...
decreased elf2 reg also seen in prion infected mice, reactivating it can rescue symptoms
repeat expansion diseases often show abberant translation produce, higher translation, these repeats form quadruplexes or hairplins


Joining the dots – protein-RNA interactions mediatinglocal mRNA translation in neurons
https://febs.onlinelibrary.wiley.com/doi/epdf/10.1002/1873-3468.13121
Galagher and Ramos et al 
Lots of dynamically localized proteins over course of develompent.
Zip codes can be structural or sequence mediated.
Zahr et al 2018 is used as example of RBP that effects translational egulation
As is FMRP's binding of a factor that sequesters the elongations factor
and regulation of polyA tails - short ones suppress translation.
FMRP binds some, not all Gquads, ref 64 might have way to find this mot
FMRP may bind ribosomal lproteins and RNAs, regulate translation
Also several short repeat motifs
ZBP1 (zipcode bindign protein 1 or IGF2 mRNA bindig protein 1 IMP1 or IGF2BP1), pre or neonatal death, synaptic morphilogy, transport, translationa, degradation of diverse mRNAs, inc beta actin, in actin the KH3,4 domain binds short motifs on eithe rof it's sides, in gengeral veyr hard to predit it's binding.
cpeb1 -transport to synapse, short polyaA maint and transport memory, binds uuuuau motif in 3' UTR - the cytoplasmic polyadenylation element.
staufen - fly embryos and neurons, and in mammals.staufen2 neurons psecific. binds dsRNA

Genome editing of upstream open reading frames enables translational control in plants
https://www.nature.com/articles/nbt.4202?WT.ec_id=NBT-201809&spMailingID=57392017&spUserID=MTc3NDMwNTgwMAS2&spJobID=1482485271&spReportId=MTQ4MjQ4NTI3MQS2
Zhang et al 2018
Targeting uORFs in plants with CRISPR targeting start codons allows modulation of some developmental genes



https://www.biorxiv.org/content/biorxiv/early/2018/08/28/401885.full.pdf
- Polioudakis1 et al
Single cell 


RNA localization is a key determinant of neurite-enriched proteome
Zappulo et al 2017
https://www.nature.com/articles/s41467-017-00690-6#ref-CR9
- note that mRNA, translation or protein can be local. 
- localization can be explained by active transport, diffusion+entrapment, or localized degredation. 
- lots of refs on mRNA localization and translation - in particular note Taliaferro et al 2016, distal final exons often mediate localization, including by presence of Mbnl1/2 motifs (musceblind) 
- We should get a lhold ofa  list of known zip codes. 
- first study to actually look at proteome rather than just mRNA (?what about riboseq?)
- Find half of neurite enriched proteome is from positioned mRNAs
- 29 neurite targetted RBPs #promising candidates for translational reg in our system.
- 1292 transcripts at least 2 fold enriched in neurite
- Find some motifs enriched, including known ZIP codes (see their other refs for other ZIP codes)

Zahr et al 2018
https://www.ncbi.nlm.nih.gov/pubmed/29395907

Cortical RPs transcriptionally primed to generate diverse types of neurons. Priming is in part implemented by the Pum2/4E-T translational repression complex, which has been shown experimetnally to regulate genes with the Pumilio motif, and disruption of which causes abberant coexperession.

https://www.biorxiv.org/content/10.1101/599373v1?rss=1&utm_source=dlvr.it&utm_medium=twitter
Magnusson et al 2019
-We should quote our median correlations to protein abundance (like they did?)


https://www.nature.com/articles/s41586-019-0888-x
Baser et al 2019 
Onset of differentiation is post-transcriptionally controlled in adult neural stem cells

> which is key to maintaining the pool of fully functional stem cells, not only in the brain but also in the bone marrow and hair follicles
Can we see stem cell genes becoming quiescent?

 > "Neurons also maintain a subset of messenger RNAs in a translationally silent state, which react ‘on demand’ to intracellular and extracellular signals."
Can we see this?

https://science.sciencemag.org/content/364/6440/eaav2522.full?ijkey=sV49inhojeB2c&keytype=ref&siteid=sci

"Temporal patterning of apical progenitors and their daughter neurons in the developing neocortex"

>Together, our findings identify a combinatorial process in which cell type–specific neuronal identity emerges from the apposition of generic differentiation programs onto embryonic age–dependent, AP-derived transcriptional states. In this scenario, neuronal differentiation essentially corresponds to the implementation of programs coding for generic neuronal features (e.g., neurites, neurotransmission) onto temporally defined initial transcriptional states. This process is reminiscent of how neuron diversity is generated in evolutionarily older brain regions such as the subpallium or spinal cord (7, 24, 42, 43). 

I'm suspicious this might be trivial. Like, generic differentiation programs have to be there. Defined states have to be there since there are defined endpoints. How could this NOT be true - if there were distinct types there originally waiting to differentiate. So are they saying that diff and temp components of the 

>In evolutionary terms, temporal patterning may have been selected as the primary mode of neuron production within neocortical areas because it allows the generation of a large spectrum of cell types at low spatial cost.

This is a non trivial, simple point, which I guess 

>In the course of corticogenesis, APs progressively become more exteroceptive. This suggests that extrinsic factors modulate neurogenesis late in development.

We see this I think.

>These environment-dependent processes may occur in an area-specific manner and may account for the recently reported molecular diversity in corresponding adult neuronal types across cortical areas (10).

## Methods
We processed data with a pipeline constructed using [Snakemake](_link), and a set of scripts written in R and Python. Code is available on github at [github link](_link). Riboseq reads were first trimmed of illumina adaptors and (2xNNNN) UMI barcodes, before being filtered for tRNA and rRNA. tRNA and rRNA was removed by aligning to tRNA and (spliced) rRNA sequences from gencode, version M12 using bowtie2. Any Riboseq read aligning to these sequences, allowing for up to 1 mismatch, was eliminated from subsequent analysis. we then used [STAR](_link) with the following parameters (insert later) randomly assigning multimapping reads to one of their best mapping positions and eliminating reads which mapped to 20 positions or more. [samtools](_link), [bamtools](_link), [Picard ](_link) and [multiqc](_link) were then used to produce an aggregated QC report for all RNAseq and RIboseq reads. [we'll want some stuff about the final results of our QC here] We then applied [____either ORFik or Riboqc___, assuming we go this route] to obtain Psite counts from our Riboseq data, and used these, along with Rsubread feature counts, to obtain gene-level RNAseq and P-site counts for our data. [Ribodiff](_link) was then used to assess differential translational efficiency. Joint modeling of riboseq and rnaseq data was carried out by transforming all data with a variance stabilizing transformation (which resulted in approximately heteroskedastic data, see figure _ ) and then fitting a linear model with [Limma](_link).

(Notes on the above - right now I'm assumign we'll sort out he periodicity issue and then get p-sites, if not we might just end up counting reads and not using ORFik or riboqc. I'm leaving out SaTAnn and the location of ORF.)




Notes, progress
Use sublime text workspaces for continuity, also use knitr, spin
Degredation model, once finished, gets used for the final edition of the clustering.
<!-- rmarkdown::render('/fast/work/groups/ag_ohler/dharnet_m/cortexomics/cortexomics.Rmd') -->

	

