GTF = 'foo.gtf'
REF = 'my_genome.fa'
bamfile = 'my_bamfile.bam'
ANNOBASE = 'myannotation'
annot_script = 'riboseqc_annotation.R'
riboqc_script = 'Ribo.seQC.R'
outfolder = 'myriboseqc'
ribqc_report = 'ribqc_report.Rmd'
sampname = 'mysampname' #this will prefix the results files generated by riboseqc

#prepare annoation
Rscript $annot_script $GTF $outfolder/${ANNOBASE}.riboqc.db $outfolder/${ANNOBASE}.annoout

#run the script - you may want to interate over multiple samples and combine them below with comma seperated lists as arguments
#Note this script needs a lot of memory - 
Rscript ${riboqc_script} ${REF} $outfolder/${ANNOBASE}.annoout ${bamfile} $outfolder/$sampname

indata = 'foo' # this should be a comma seperated list of the results files produced by the script above - they have 'results_all_toknit' in their name
sinputs = 'foo' # This should be a comma seperated list of names which were passed to the above script
figpath = 'myfiguredirectory' #where you want the figures printed
reportfile = 'myreport.html' #the actual report file

#note I use an R one liner for this - you may want to open up R and look at markdown::render, and the report file itself, to understand hwat's going on here
R -e 'rmarkdown::render("'$ribqc_report'",params = list(input_list = "'$indata'",input_list_names = "'$sinputs'", output_fig_path = "'$figpath'"),output_file = "'$reportfile'")'