---
title: "Marker Gene Report"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,root.dir = "/home/dharnett/cortexomics/" )
library(rtracklayer)
library(tidyverse)
library(stringr)
library(magrittr)
library(assertthat)

select =dplyr::select
```


## Our Marker genes
We want an expression profile from RNAseq across development, for Riboseq and Total Rnaseq, for these genes:

1. PA2G4 (EBP1), both isoforms there are two isoforms in mouse - long ("p48") vs. short ("p42") - differing in the N-terminus
see attached about the isoforms' sequences

2. Pax  shows timing of stem cell expression timing

3. Tle4 shows timing of lower layer neuron differentiation

4. Bcl11b (Ctip2) shows timing of lower-middle layer neuron differentiation

5. Satb2 shows timing of upper layer neuron differentiation 

6. Cdp (Cux1) also shows timing of upper layer neuron differentiation

To do this we'll need to...

1. Load up information linking transcripts to genes - I manually used gencode IDs to be safe.

2. Get normalized expression values (TPM - transcripts per million) for our transcripts. I did this using Kallisto. The GENCODE M12 annotation release was used, with the mm10 mouse genome.

3. Sum the expression values for most genes, keeping our isoforms seperate for PA2G4

4. Calculate translational efficiency as the ratio of the Total RNAseq reads to the Riboseq reads.

5. Finally plot these values over our developmental timepoints.



```{r load_data, include=FALSE,cache = TRUE}

kallistodir = "~/cortexomics/data/kallisto_out/"
metacols = c('time','assay','replicate')
#create table for 
sep_iso_genes <- c('ENSMUSG00000025364','ENSMUSG00000020484')

setwd(root)

kallistocounts<-read_tsv(file.path(root,'exploration/tables/kallistocounts.tsv'),col_types=cols(
  transcript_id = col_character(),
  signal = col_double(),
  dataset = col_character(),
  time = col_character(),
  assay = col_character(),
  rep = col_integer(),
  gene_id = col_character(),
  gene_name = col_character()
))

#get our counts out of kallisto
stopifnot(exists('kallistocounts'))

#list of genes we want to plot
markergenestbl <- readr::read_tsv(file.path(root,"ext_data/marker_genes.tsv"),col_names = TRUE)
markergenes <- markergenestbl$gencode %>% str_split(',') %>% flatten_chr%>%str_replace('\\.\\d+','')

#get the correct genes
annotation_gr$stripped_geneID <-  str_replace(annotation_gr$geneID,'\\.\\d+$','')
markergene_gr <- annotation_gr%>% subset( (.$type=='transcript') & (stripped_geneID %in% markergenes) )
stopifnot(all(markergenes %in% markergene_gr$stripped_geneID))

#get counts for specific genes
# markergene_counts <- markergene_gr%>%mcols%>%as.data.frame%>%select(ID,gene_name)%>%left_join(htseqcounts)
kallistocounts <- kallistocounts %>% mutate(ID = str_extract(transcript_id,'ENSMUST\\d+'))
markertranscript_tpm <- kallistocounts%>%filter(ID)
markertranscript_tpm%>%head

#this table now contains the transcripts and genes requested
marker_tpm <- bind_rows(
  markertranscript_tpm%>%subset(gene_id%in%sep_iso_genes)%>%select(name=Name,dataset,tpm=signal),
  markertranscript_tpm%>%subset(! gene_id%in%sep_iso_genes)%>%group_by(gene_id,dataset,gene_name)%>%summarise(tpm = sum(signal))%>%select(ID=gene_id,name=gene_name,dataset,tpm)
)



marker_tpm <- marker_tpm %>% rename(TPM = tpm)

#metadata for our datasets
marker_tpm <- 
  marker_tpm%>%
	mutate(tmp = dataset%>%str_replace('_28_30','-28-30'))%>%
	separate(tmp,metacols,sep='_')%>%
  mutate(assay = str_replace(assay,'-28-30',''))

#calculate translational efficiency
markerTE <-  marker_tpm%>%
  select(assay,name,time,replicate,TPM)%>%
  spread(assay,TPM)%>%
  mutate(TE = ribo/total)%>%select(-ribo,-total)

markerTE<-
    markerTE%>%
    group_by(name)%>%mutate(TE = log2(TE/min(TE[time=='E13'])))%>%
    mutate(TE =  replace(TE,is_in(TE,c(Inf)),NA))

#summarise tpms by their 
# marker_tpm <- marker_tpm %>%group_by(name,time,assay)%>%  summarise(TPM = sum(TPM))
#check tbls look right
marker_tpm %>% has_name(c('TPM','time','name','assay')) %>% all %>% assert_that(msg = 'names missing')
markerTE %>% has_name(c('TE','time','name')) %>% all %>% assert_that(msg = 'names missing')

'ENSMUST00000063084'
'ENSMUST00000149623'

```
## Total And Ribocapture RNAseq read TPM (Transcripts Per Million) for our marker genes:
(dots represent individual replicate values)

```{r TPMplots, echo=FALSE,cache = TRUE}
names = unique(marker_tpm$name)
names %<>% grep(x=.,v=TRUE,invert=TRUE,'-003')

for(name_i in names){
marker_tpm %>% filter(name==name_i)%>%
  mutate(TPM = TPM+0.001)%>%
  { qplot(data=.,y=TPM,x = time, geom = 'point' )+
      facet_grid( scale = 'free',assay ~ isoform )+
      #scale_y_log10(breaks=10^(0:3),limits=c(1,1e3))+
      scale_y_continuous(limits = c(0,200))+
      expand_limits(y=0)+
      ggtitle(str_interp("TPM for ${name_i}"))+
      theme_bw()+
      theme(text = element_text(size = 16))
  }%>%print
}


```

## Translational Efficiency (RiboTotal) per timepoint/replicate
(replicates divided seperately)

Note that in some cases the total counts are very low - this will render the TE - which is a ratio, unreliable.

I'm also unsure if transcript specific TE can be trusted - there may be differences in the ability to attribute total reads and Riboseq reads to their parent transcript...

Note that values are not displayed if the TotalSeq reads are zero.

```{r te, echo=FALSE,cache = FALSE}
names = unique(markerTE$name)
names %<>% grep(x=.,v=TRUE,invert=TRUE,'-003')

for(name_i in names){
                  

markerTE %>% filter(name==name_i)%>%
  { 
  ymax = .[['TE']]%>%keep(is.finite)%>%keep(Negate(is.na))%>%keep(Negate(is.nan))%>%max%>%multiply_by(1.2)

    message(ymax)
        qplot(data=.,y=`TE`,x = time, geom = 'point' )+
      #scale_y_log10(breaks=10^(0:3),limits=c(1,1e3))+
      scale_y_continuous(limits = c(-1,ymax),name = 'TE - Log2(T/T0)')+
      ggtitle(str_interp("Translational Efficiency for ${name_i}"))+
      theme_bw()+
      theme(text = element_text(size = 16))
  }%>%print
  
}

```


