---
title: "Marker Gene Report"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,root.dir = "/home/dharnett/cortexomics/" )
library(rtracklayer)
library(tidyverse)
library(stringr)
library(magrittr)
group_slice<-function(dt,v){
  #subset our original data
  dt[group_indices(dt) %in% v , ]
}
select =dplyr::select
```


## Our Marker genes
We want an expression profile from RNAseq across development, for Riboseq and Total Rnaseq, for these genes:

1. PA2G4 (EBP1), both isoforms there are two isoforms in mouse - long ("p48") vs. short ("p42") - differing in the N-terminus
see attached about the isoforms' sequences

2. Pax  shows timing of stem cell expression timing

3. Tle4 shows timing of lower layer neuron differentiation

4. Bcl11b (Ctip2) shows timing of lower-middle layer neuron differentiation

5. Satb2 shows timing of upper layer neuron differentiation 

6. Cdp (Cux1) also shows timing of upper layer neuron differentiation

To do this we'll need to...

1. Load up information linking transcripts to genes - I manually using gencode IDs to be safe 

2. get the counts for various transcripts linked to our genes. These have been obtained using the fastq's that Ulrike provided, and the BIH pipeline, which maps reads using STAR, and HTseq, which produces feature counts from mapped reads and annotation files. The GENCODE M12 annotation release was used, with the mm10 mouse genome.

3. We then transform our counts to Tags Per Million (TPM) values by counting mapped reads per library.

4. Finally we plot these values over our developmental timepoints.



```{r load_data, include=FALSE,cache = TRUE}

setwd("/home/dharnett/cortexomics/")
kallistodir = "~/cortexomics/data/kallisto_out/"
keepisogenes = 'Pa2g4'

# 

#list of genes we want to plot
markergenestbl <- readr::read_tsv("~/cortexomics/data/marker_genes.tsv",col_names = TRUE)
markergenes <- markergenestbl$gencode %>% str_split(',') %>% flatten_chr
#get list of genes in genera
annotation_gr <- rtracklayer::import("~/bih_cluster/projects/cubit/current/static_data/annotation/GENCODE/M12/GRCm38/gencode.vM12.annotation.gff3")


#get the correct genes
annotation_gr$stripped_ID <- str_replace(annotation_gr$gene_id,'\\.\\d+$','')
markergene_gr <- annotation_gr%>% .[(.$type=='transcript') & (.$stripped_ID %in% markergenes)]
stopifnot(all(markergenes %in% markergene_gr$stripped_ID))

#kallisto counts
kallistocounts <- 
  kallistodir %>%
  list.files(rec=TRUE,full=TRUE, patt="abundance.tsv")%>%
  setNames(.,basename(dirname(.)))%>%
  map(read_tsv)
kallistocounts%<>%bind_rows(.id='dataset')
kallistocounts$dataset%>%unique

#get counts for specific genes
kallistocounts%<>% mutate(transcript_id = target_id%>%str_extract('ENSMUST[\\.\\d]+'))
kallistocounts%<>% mutate(gene_id = target_id%>%str_extract('ENSMUSG[\\.\\d]+'))
kallistocounts%>%group_by(target_id,dataset)%>%filter(n()>1)%>%group_slice(1)

markergene_counts <- markergene_gr%>%mcols%>%as.data.frame%>%select(ID,gene_name,transcript_name)%>%left_join(kallistocounts%>%select(-target_id),by=c('ID'='transcript_id'))

#metadata for our datasets
group_df <- markergene_counts$dataset%>%unique%>%cbind(.,str_split_fixed(.,'_',3))%>%as.data.frame%>%set_colnames(c('dataset','time','assay','replicate'))
#add in data on time, assay etc.
group_df$dataset %<>% as.character
markergene_counts <- markergene_counts %>% left_join(group_df,by='dataset')


#split things we want to preserve isoforms for from those that dont
# markergene_counts <-
markergene_counts %<>%  mutate(ID = ifelse(gene_name %in% keepisogenes,ID,gene_id))
markergene_counts %<>%  mutate(name = ifelse(gene_name %in% keepisogenes,transcript_name,gene_name))

#sum the isoforms if we're not treating sep
markergene_counts %<>%  group_by(assay,ID,name,time,replicate) %>% summarize (tpm = sum(tpm))

countsTE <-
  markergene_counts%>%
  select(assay,ID,name,time,replicate,tpm)%>%
  spread(assay,tpm)%>%
  mutate(`Ribo/Total` = ribo/total)%>%select(-ribo,-total)
# 
# markergene_counts <-
#   markergene_counts%>%
#   group_by(gene_name,ID,time,assay)%>%
#   summarize(tpm = mean(tpm))
# 


```
## Total And Ribocapture RNAseq read counts for our marker genes:
(dots represent individual replicate values)

```{r counts, echo=FALSE,cache = TRUE}
gene_names = unique(markergene_counts$name)

for(gene_name_i in gene_names){

  markergene_counts %>% filter(name==gene_name_i)%>%
  mutate(tpm = tpm+0.001)%>%
  { qplot(data=.,y=tpm,x = time, geom = 'point' )+
      facet_grid( scale = 'free',assay ~ . )+
      #scale_y_log10(breaks=10^(0:3),limits=c(1,1e3))+
      scale_y_continuous()+
      expand_limits(y=0)+
      ggtitle(str_interp("TPM for ${gene_name_i}"))+
      theme_bw()+s
      theme(text = element_text(size = 16))
  }%>%print
  
  }


```

## Translational Efficiency (RiboTotal) per timepoint/replicate
(replicates divided seperately)

```{r te, echo=FALSE,cache = TRUE}
gene_names = unique(markergene_counts$gene_name)


for(gene_name_i in gene_names){
countsTE %>% filter(gene_name==gene_name_i)%>%
  { qplot(data=.,y=`Ribo/Total`,x = time, geom = 'point' )+
      #scale_y_log10(breaks=10^(0:3),limits=c(1,1e3))+
      scale_y_continuous()+
      expand_limits(y=0)+
      ggtitle(str_interp("Translational Efficiency for ${gene_name_i}"))+
      theme_bw()+
      theme(text = element_text(size = 16))
  }%>%print
}

```


